Automatically make PRs to bump external stdlibs

DilumAluthge opened the pull request:
Fixes #34413 

Note: this bot will make pull requests to bump external stdlibs, but it will not merge them. The pull requests will need to be manually merged by a human. This is intentional - it allows a human to look look at the pull request CI results before merging it.

Here are some examples of the pull requests that this bot will create. As you can see, the bot prints the log of all commits between the old commit and the new commit.
- `Bump Pkg to 831c11a`: https://github.com/aluthge-test/julia/pull/26
- `Bump Statistics to da6057b`: https://github.com/aluthge-test/julia/pull/27

DilumAluthge committed:
Set up the BumpStdlibs bot, and run the cron job one time per week

fredrikekre commented on .github/workflows/bump_stdlibs.yml:
```diff
@@ -0,0 +1,25 @@
+
+on:
+  schedule:
+    - cron: '00 * * * *'
```
How about weekly instead? Or every second/third day or something?

DilumAluthge replied:
I've changed it to run three times a week (00:00 UTC on Mondays, Wednesdays, and Fridays). How does that sound?

I guess we could make it twice a week if you want? One time per week seems a little too infrequent - Pkg can certainly accumulate a lot of new commits in seven days.

KristofferC replied:
> One time per week seems a little too infrequent - Pkg can certainly accumulate a lot of new commits in seven days.

I think one time a week is very much sufficient.

fredrikekre replied:
Yea, if there are urgent fixes one can manually update, this is mostly just so we don't forget to do it at all during a month etc. I think it will quickly be annoying if we get bump PRs too often.

DilumAluthge replied:
Sounds good. I’ve changed it to run once per week (00:00 UTC every Monday)

DilumAluthge commented:
I'm pretty sure that the CI failure on `buildbot/tester_linux32` is unrelated. Can someone with the appropriate permissions re-trigger the `buildbot/tester_linux32` job?

vtjnash commented:
Wow, this seems quite fancy. I was picturing someone would just write something like the `contrib/refresh_bb_tarballs.sh` to manually do this.

I think it should probably be getting the release (and version) data from https://github.com/JuliaRegistries/General though, and not literally following master.

DilumAluthge commented:
> I think it should probably be getting the release (and version) data from https://github.com/JuliaRegistries/General though, and not literally following master.

I’m not sure what you mean by this. Pkg and Statistics are not in the General registry.

StefanKarpinski commented:
Very cool. Can you say a bit at a high level about what this is doing? Why is so much machinery necessary? Is it just to get the nice PR body listing the contents or is there some reason that filtering through all the PRs is needed to figure out what the update is?

ararslan commented:
The code part should go in `contrib` rather than `stdlib`, IMO, especially since it relies on non-stdlib packages.

DilumAluthge commented:
> The code part should go in `contrib` rather than `stdlib`, IMO, especially since it relies on non-stdlib packages.

I agree. I've moved the file `BumpStdlibs.jl` to the `contrib` directory.

DilumAluthge commented:
> Is it just to get the nice PR body listing the contents

That is correct. To emphasize this, I have refactored that code out into its own function, and I've added some comments to clarify.

DilumAluthge commented:
Bump. Any additional thoughts on this?

DilumAluthge commented on .github/workflows/bump_stdlibs.yml:
```diff
@@ -0,0 +1,25 @@
+
+on:
+  schedule:
+    - cron: '0 0 * * 1'
```
Yep! Can be modified of course.

staticfloat replied:
That seems totally reasonable to me.  I don't think we really want this opening PRs more often than that. ;)

vtjnash replied:
This feels a bit like the wrong design. We would want something like this to trigger on a package making a release, not on a time schedule. Setting `*_BRANCH = master` is usually only used with git-external when an external dependency is young, and not yet stable enough to make releases. But usually, it should be set to a real branch (or preferably probably a tag).

DilumAluthge commented on .github/workflows/bump_stdlibs.yml:
```diff
@@ -0,0 +1,25 @@
+      - name: Update external stdlibs
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: julia -e 'include("contrib/BumpStdlibs.jl"); "${{ github.repository }}" == "JuliaLang/julia" && BumpStdlibs.main("${{ github.repository }}", ["Pkg" => "https://github.com/JuliaLang/Pkg.jl", "Statistics" => "https://github.com/JuliaLang/Statistics.jl"]; cc_usernames = String["@DilumAluthge"])'
```
Yeah, maybe a TOML file that we read in?

staticfloat replied:
Or a multiline string that we feed in to the `julia` process or something? I believe these yml files have a nice block syntax that we can use for multiline input.

DilumAluthge replied:
Alright @staticfloat take a look. I split the long line into multiline input.

DilumAluthge commented on contrib/BumpStdlibs.jl:
```diff
@@ -0,0 +1,365 @@
+@inline function always_assert(cond::Bool, msg::String)::Nothing
+    cond || throw(AlwaysAssertionError(msg))
+    return nothing
+end
```
Yeah I can change this. There must have been a reason in the original (most of this code is from CompatHelper), but it doesn’t seem necessary here.

DilumAluthge replied:
Done.

DilumAluthge commented on contrib/BumpStdlibs.jl:
```diff
@@ -0,0 +1,365 @@
+
+                # @info("Running `make` inside the `deps` subdirectory")
+                # cd(joinpath(main_repo_dir, "deps"))
+                # run(`make`)
```
Yeah this is not necessary for Pkg or Statistics which is why I commented it out. But I figured it might be necessary in the future.

staticfloat commented:
This isn't something to do now, but if so much of this code is more-or-less boilerplate (finding other pull requests that were opened by this bot, updating its contents, etc...) we should probably factor that out into a reusable package that can just be used by this script.  Conceptually this is very simple, and I'm 100% certain those building blocks would be used by other packages.  This is really cool stuff @DilumAluthge!

DilumAluthge commented:
> we should probably factor that out into a reusable package that can just be used by this script

Yeah that definitely seems like a good idea. I'm thinking that we leave it in this script at first, let it run for a while, and iron out all the bugs. Then once it looks pretty stable we can put it in its own package.

DilumAluthge committed:
Fix a typo

DilumAluthge committed:
Fix another typo in the workflow file

KristofferC commented:
Personally I feel that the fact that this does stuff automatic via github actions contributes to a lot of complexity. I wonder if a script that one runs locally to update the stdlib branches is enough, and then one can manually push that branch (similar to how the backport script works).

StefanKarpinski commented:
Hard for triage to review, but let's just try it and see how it goes.

JeffBezanson commented:
Testing this is a bit of a concern. Will be interesting to see if / how quickly this bitrots, and if so we'll have to put some kind of tests in place. But we can try it and see what happens.

KristofferC commented on .github/workflows/bump_stdlibs.yml:
```diff
@@ -0,0 +1,31 @@
+        with:
+          version: ${{ matrix.julia-version }}
+      - name: Install dependencies
+        run: julia -e 'import Pkg; Pkg.add(["GitCommand", "GitHub"])'
```
Should use a manifest and instantiate so this uses constant versions. 

Also why is  GitCommand needed? I think requiring that git is installed is totally fine.

DilumAluthge replied:
Good idea, I’ll add a manifest.

I can also remove GitCommand and use local git instead.

DilumAluthge commented:
The plan is to develop/prototype this in https://github.com/JuliaPackaging/BumpStdlibs.jl.
